### PR TITLE
Add ordered dictionary for quicker look up of Installation packages in StatusHandler 

### PR DESCRIPTION
**Pull request by feng-j678**
What is in this PR:
[X] Modify the nested for loop in set_package_install_status and set_package_install_status_classification to use order dictionary for quicker look up
[X] unit test installation order dictionary

### CODE REVIEW

**Review by kjohn-msft — DISMISSED** [review]

File: src/core/src/service_interfaces/StatusHandler.py [hunk 1]
```diff
@@ -159,15 +161,19 @@ def set_package_install_status(self, package_names, package_versions, status="Pe
...
-                if patch_id == self.__installation_packages[i]['patchId']:
-                    patch_already_saved = True
-                    if classification is not None:
-                        self.__installation_packages[i]['classifications'] = [classification]
-                    self.__installation_packages[i]['patchInstallationState'] = status
+            # Match patch_id in map and update existing patch's classification i.e from None -> security and update pending status
+            if len(self.__installation_packages_map) > 0 and patch_id in self.__installation_packages_map:
+                if classification is not None:
+                    self.__installation_packages_map.setdefault(patch_id, {})['classifications'] = [classification]
+                self.__installation_packages_map.setdefault(patch_id, {})['patchInstallationState'] = status
```
Comment (rane-rajasi) [inline src/core/src/service_interfaces/StatusHandler.py:175]: L 174 is marked is marked to be missing code coverage

File: src/core/src/service_interfaces/StatusHandler.py [hunk 2]
```diff
@@ -206,18 +214,24 @@ def set_package_install_status_classification(self, package_names, package_versi
 
         self.validate_packages_being_installed(package_names, package_versions)
 
+        # lazy initialize OrderedMap, problem can't initialize any objects in __init__
+        if self.__installation_packages_map is None:
+            self.__installation_packages_map = OrderedDict()
+
```
Comment (rane-rajasi) [inline src/core/src/service_interfaces/StatusHandler.py:220]: L 219 not covered by tests

File: src/core/tests/Test_StatusHandler.py [hunk 3]
```diff
@@ -395,7 +395,45 @@ def test_sort_packages_by_classification_and_state(self):
...
 
+        status_handler = StatusHandler(self.runtime.env_layer, self.runtime.execution_config, self.runtime.composite_logger, self.runtime.telemetry_writer, self.runtime.vm_cloud_type)
+        self.runtime.execution_config.operation = Constants.INSTALLATION
+        self.runtime.status_handler.set_current_operation(Constants.INSTALLATION)
+
+        patch_count_for_test = 50
+        test_packages, test_package_versions = self.__set_up_packages_func(patch_count_for_test)
+
+        status_handler.set_package_install_status(test_packages, test_package_versions, 'Installed', 'Other')
+        self.assertIsNotNone(status_handler._StatusHandler__installation_packages_map)
```
Comment (rane-rajasi) [inline src/core/tests/Test_StatusHandler.py:416]: Same comment as the other PR for OrderedDict in assessment. Tests should check a function's output based on different inputs. We should never access the internal variables within a function

**Review by rane-rajasi — CHANGES_REQUESTED** [review]
Comments inline

File: src/core/tests/Test_StatusHandler.py (hunk 3 above)
Comment (rane-rajasi) [reply src/core/tests/Test_StatusHandler.py:416]: Adding @kjohn-msft to get this inputs

File: src/core/src/service_interfaces/StatusHandler.py (hunk 2 above)
Comment (feng-j678) [reply src/core/src/service_interfaces/StatusHandler.py]: move logic to __init__ and test cover

File: src/core/tests/Test_StatusHandler.py (hunk 3 above)
Comment (feng-j678) [reply src/core/tests/Test_StatusHandler.py:425]: add more assertion and test the load_status_file to assert the package_map is working as expected

File: src/core/src/service_interfaces/StatusHandler.py (hunk 1 above)
Comment (feng-j678) [reply src/core/src/service_interfaces/StatusHandler.py:172]: move logic to init and test cover

File: src/core/tests/Test_StatusHandler.py (hunk 3 above)
Comment (rane-rajasi) [reply src/core/tests/Test_StatusHandler.py:425]: Same comment as the assessment PR:

We don't need to check StatusHandler's internal variables to verify if packages exist. Line 418 is reading contents from the status file, you can directly check if the packages exist within that file.
Please remove lines 424 to 429 and check for packages within status file using the substatus_file_data in line 412.

Please see lines 420 to 423 for reference.

File: src/core/tests/Test_StatusHandler.py [hunk 4]
```diff
@@ -395,7 +396,114 @@ def test_sort_packages_by_classification_and_state(self):
...
+        formatted_message = json.loads(substatus_file_data['formattedMessage']['message'])
+        self.assertEqual(len(formatted_message['patches']), patch_count_for_test)
+        self.assertEqual(formatted_message['patches'][0]['classifications'], ['Critical'])
+        self.assertEqual(formatted_message['patches'][0], expected_value_critical)
+        self.assertEqual(len(status_handler._StatusHandler__installation_packages), patch_count_for_test)
+        self.assertIsNotNone(status_handler._StatusHandler__installation_packages_map)
+        self.assertEqual(status_handler._StatusHandler__installation_packages_map[patch_id_critical]['name'], 'python-samba0')
+        self.assertEqual(status_handler._StatusHandler__installation_packages_map[patch_id_critical]['classifications'], ['Critical'])
+        self.assertEqual(status_handler._StatusHandler__installation_packages_map[patch_id_critical], expected_value_critical)
+        self.assertEqual(len(status_handler._StatusHandler__installation_packages_map), patch_count_for_test)
```
Comment (rane-rajasi) [inline src/core/tests/Test_StatusHandler.py:445]: Same comment applies about not using a class'es internal variables in asserts

File: src/core/tests/Test_StatusHandler.py [hunk 5]
```diff
@@ -395,7 +396,114 @@ def test_sort_packages_by_classification_and_state(self):
...
+
+    def test_load_status_and_set_package_install_status(self):
```
Comment (rane-rajasi) [inline src/core/tests/Test_StatusHandler.py:447]: Isn't test_installation_packages_map() also checking package install status? Why do we need this test again?

File: src/core/tests/Test_StatusHandler.py [hunk 6]
```diff
@@ -395,7 +396,114 @@ def test_sort_packages_by_classification_and_state(self):
...
+        self.assertEqual(json.loads(substatus_file_data["formattedMessage"]["message"])["patches"][0]["name"], "python-samba0")
+        self.assertTrue('Critical' in str(json.loads(substatus_file_data["formattedMessage"]["message"])["patches"][0]["classifications"]))
+        self.assertEqual(json.loads(substatus_file_data["formattedMessage"]["message"])["patches"][1]["name"], "python-samba1")
+        self.assertEqual('Critical', str(json.loads(substatus_file_data["formattedMessage"]["message"])["patches"][1]["classifications"][0]))
+        self.assertEqual(json.loads(substatus_file_data["formattedMessage"]["message"])["patches"][2]["name"], "python-samba2")
+        self.assertEqual('python-samba0_2:4.4.5+dfsg-2ubuntu5.4_Ubuntu_16.04', str(json.loads(substatus_file_data["formattedMessage"]["message"])["patches"][0]["patchId"]))
+        self.assertTrue('Critical' in str(json.loads(substatus_file_data["formattedMessage"]["message"])["patches"][2]["classifications"]))
+        self.runtime.env_layer.file_system.delete_files_from_dir(self.runtime.status_handler.status_file_path, '*.complete.status')
+
+    def test_load_status_and_set_package_install_status_classification(self):
```
Comment (rane-rajasi) [inline src/core/tests/Test_StatusHandler.py:472]: Same question as the one for test_load_status_and_set_package_install_status()

**Review by rane-rajasi — CHANGES_REQUESTED** [review]
Comments inline

File: src/core/tests/Test_StatusHandler.py (hunk 6 above)
Comment (feng-j678) [reply src/core/tests/Test_StatusHandler.py:462]: test_load_status_and_set_package_install_status is checking if we update existing map object classifications and state, and this test test_load_status_and_set_package_install_status_classification is just checking classification, it's an extra test to cover line 220 - 222

File: src/core/tests/Test_StatusHandler.py (hunk 5 above)
Comment (feng-j678) [reply src/core/tests/Test_StatusHandler.py:437]: it's different test_installation_packages_map() is for checking if the values are same on final output, the test_load_status_and_set_package_install_status is checking if we update existing map object classifications and state, so to test this we need to first need to load complete data from status folder, then perform update on them

File: src/core/tests/Test_StatusHandler.py (hunk 4 above)
Comment (feng-j678) [reply src/core/tests/Test_StatusHandler.py]: internal variable is removed

**Review by kjohn-msft — DISMISSED** [review]

**Review by rane-rajasi — DISMISSED** [review]
@feng-j678 Now that the py27 CI issue is fixed, let that PR go in first, so you can fetch latest from master into this PR, run py27 on this PR and then merge.

**Review by rane-rajasi — APPROVED** [review]

**Review by kjohn-msft — APPROVED** [review]

**Pull request merged**